### DISCUSSION TITLE
Using `INSERT INTO ... SELECT` with Session

### DISCUSSION THREAD

**Discussion by luciengaitskell** [sqlalchemy/sqlalchemy#7894] [Usage Questions] [answered]
Is it possible to configure a `session.add(...)` to use `INSERT INTO ... SELECT` instead of `INSERT INTO ... VALUES`?

I am using Snowflake, and would like to insert JSON as part of a row. I've added a `TypeDecorator` to wrap the bind value with `func.parse_json(...)`. However, functions are not allowed as part of `VALUES`.

I'd like to be able to indicate to SQL Alchemy, that when inserting it must be performed with a `INSERT INTO ... SELECT`.

**Comment by zzzeek** [comment, marked as answer]
> Is it possible to configure a `session.add(...)` to use `INSERT INTO ... SELECT` instead of `INSERT INTO ... VALUES`?

no, and this would not really make much sense.

> 
> I am using Snowflake, and would like to insert JSON as part of a row. I've added a `TypeDecorator` to wrap the bind value with `func.parse_json(...)`.  However, functions are not allowed as part of `VALUES`.


Use the JSON type at https://docs.sqlalchemy.org/en/14/core/custom_types.html#marshal-json-strings .  There is no need to use server-side functions to persist or load JSON values, JSON values are marshalled into strings before being passed to the database driver.

**Reply by luciengaitskell** [reply]
@zzzeek I understand, and I'd prefer to load data that way. This worked great for Postgres. Snowflake, however, [does not support that syntax](https://docs.snowflake.com/en/sql-reference/data-types-semistructured.html#example-of-inserting-an-object).

Do you have any suggestions on how to work around this? One option is to just forgo the ORM when adding JSON data.

**Reply by zzzeek** [reply]
> @zzzeek I understand, and I'd prefer to load data that way. This worked great for Postgres. Snowflake, however, [does not support that syntax](https://docs.snowflake.com/en/sql-reference/data-types-semistructured.html#example-of-inserting-an-object).

I'm not really familiar with Snowflake but took a look at their documentation and this appears to be a fully featured SQL database, with support INSERT..VALUES right here:  https://docs.snowflake.com/en/sql-reference/sql/insert.html

if you use the VARCHAR oriented JSON datatype I indicated above, there is no issue with using a regular INSERT..VALUES because the database is not dealing with JSON, you're inserting a string.

however it appears that snowflake has a JSON object type, and perhaps you are using that datatype as it mentions the syntax you're looking for at https://docs.snowflake.com/en/sql-reference/sql/insert.html#usage-notes.   If you'd like this syntax to be transparently supported by "insert()", such that the "insert" statements emitted by the unit of work can take advantage of it, I would [file an issue with the snowflake dialect](https://github.com/snowflakedb/snowflake-sqlalchemy/issues).

**Reply by zzzeek** [reply]
here's the issue which addresses this:  https://github.com/snowflakedb/snowflake-sqlalchemy/issues/107

**Reply by zzzeek** [reply]
just as a note, you **can** make this happen, it just would require writing more code.  I would approach it by writing an [insert compilation handler](https://docs.sqlalchemy.org/en/14/core/compiler.html#changing-the-default-compilation-of-existing-constructs) which receives the textual "INSERT" statement from the usual compiler, then does a regular expression search in VALUES and rewrites it in an exactly equvialent "INSERT..SELECT" format.  That assumes however that the INSERT..SELECT format can accommodate all the same features as an INSERT..VALUES on this particular database.

**Reply by luciengaitskell** [reply]
@zzzeek Thanks for the advice here. I was able to successfully build a custom compiler function for `visit_insert` to convert to a `insert.from_select` when a JSON column is selected. This works properly with the `insert()` core operations, but I'm still having problem with the ORM working quite as I'd like.

The ORM seems to select the specific columns in the `INSERT` before executing on the DB, but this appears to occur after the `Insert` object hits the compiler. Thus, when `visit_insert` is run, the `.values(...)` method has not been run and no selected columns are provided. This shouldn't necessarily be a problem, as I can just use the column list for the table and use each column's default value configured in the schema. However, I'm having a compatibility issue between how Snowflake wants me to configure [auto-incrementing columns](https://docs.snowflake.com/en/user-guide/sqlalchemy.html#auto-increment-behavior) and how SQLALchemy believes my primary key column is empty (likely because it isn't provided in the ORM object). I'd prefer to just not pass any values for the empty columns.

One fix would be to only include the columns with specified values in the `INSERT ... SELECT`, however, I don't seem to have this information when the `visit_insert` is executed. Is there another compiler function that may be run instead during this process in the ORM?

**Reply by zzzeek** [reply]
the compiler is not necessarily using insert().values(), it's using insert() by itself, but when the compiler receives the statement there is a [column_keys](https://github.com/sqlalchemy/sqlalchemy/blob/c00d653a4bff82ec4765d2bfba4ed19f1e201ae2/lib/sqlalchemy/sql/compiler.py#L706) collection that indicates the columns that are present in the parameter list which were sent to connection.execute().  the visit_insert() routine then goes through quite a complicated process to figure out exactly how to render the desired INSERT statement.  Hence my suggestion to let the compiler do all that work, receive the string, then simply replace VALUES with SELECT.

Here is a complete front to back demo using SQLite

```py
import json
import re

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import FetchedValue
from sqlalchemy import func
from sqlalchemy import Insert
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.sql import Insert
from sqlalchemy.types import UserDefinedType

Base = declarative_base()


class SnowflakeJson(UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "JSON"

    def bind_processor(self, dialect):
        def process(value):
            if value is not None:
                value = json.dumps(value)
            return value

        return process

    def result_processor(self, dialect, coltype):
        def process(value):
            if value is not None:
                value = json.loads(value)
            return value

        return process

    def bind_expression(self, bindvalue):
        # use this for snowflake
        # return func.parse_json(bindvalue)

        # for our SQLite demo, use the SQLite json() function just
        # to illustrate.  NULL handling may need to use a function like
        # NVL or a CASE statement depending on snowflake's quirks
        return func.json(bindvalue)


@compiles(Insert)
def _rewrite_insert(stmt, compiler, **kw):
    insert_sql = compiler.visit_insert(stmt, **kw)

    # look in the columns being inserted, see if there's
    # JSON being inserted.  also can just look at the INSERT string
    # and look for the json function
    for key in compiler.column_keys:
        if isinstance(stmt.table.c[key].type, SnowflakeJson):
            use_json = True
            break
    else:
        use_json = False

    if not use_json:
        return insert_sql

    stmt_reg = re.match(
        r"^INSERT INTO (.+?) \((.+?)\) VALUES \((.+)\)$", insert_sql
    )
    if not stmt_reg:
        return insert_sql

    else:
        # rewrite INSERT as per
        # https://docs.snowflake.com/en/sql-reference/sql/insert.html#usage-notes
        return (
            f"INSERT INTO {stmt_reg.group(1)} "
            f"({stmt_reg.group(2)}) SELECT {stmt_reg.group(3)}"
        )


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)

    # the "empty" server_default will cause the ORM to not include the
    # NULL column in the INSERT statement when no value is present.
    json_data = Column(SnowflakeJson, server_default=FetchedValue())


# will be our "snowflake" for this exercise
e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(data="no JSON"))

# emits INSERT..VALUES
s.flush()

s.add(A(data="ok this has JSON", json_data={"foo": "bar"}))

# emits INSERT..SELECT
s.flush()
```

**Reply by luciengaitskell** [reply]
@zzzeek Thank you for the code snippet. That works really well 👍 

Using `column_keys`, I was able to get quite far with my initial approach. The one problem is that Snowflake does not supporting returning values with an `INSERT`. This prevented the ORM from receiving the ID of the inserted row. The ORM handles this in the standard `INSERT ... VALUES` by executing a `nextval` on the sequence separately, before executing the `INSERT`. However, the ORM does not seem to do this with `Insert.from_select` objects. Is there a way to indicate or execute some pre-insert command, just like the ORM does automatically with `INSERT ... VALUES`?

I can open up a new discussion post if necessary. Thanks!

**Reply by zzzeek** [reply]
Right, the nextval() thing will not happen if you are passing SQLAlchemy Core an Insert().from_select(), because this is what's known as an "inline" insert() statement.   the from_select() can deliver any number of rows, so there's no way to call "nextval()" separately in this case.

if you do string translation as above, SQLAlchemy will still consider the INSERT to be a single row, non-inline statement and it can invoke the nextval().    if you are having the compiler process a real insert().from_select() instead of the statement that was given, then you would need to manipulate other internal parts of the compiler's state so that it still runs the nextval().   I could show you how to do this but it's API that's not really something people normally use on the outside.     I think simple string translate is a lot safer here.

**Reply by luciengaitskell** [reply]
Fantastic, I really appreciate the explanation. I agree- for this application string translation works very well and I'll stick with that.

Out of interest, do you have some links or examples for the compiler state modification you mention above? It's been quite helpful digging into the SQLA internals to properly understand how to use it.

**Reply by zzzeek** [reply]
the docs for the compiler extension are at https://docs.sqlalchemy.org/en/14/core/compiler.html .    there could be more examples but there is enough to get people started.

**Reply by luciengaitskell** [reply]
Thanks, that documentation has been quite useful.

There's just one final thing that's just not quite clear to me- How would I be able to implement an operation that executes an additional command, such as the `SELECT ...nextval` before `INSERT`. That would, for example, let me extend the `Insert` compiler to also support `from_select()` with a sequence. I couldn't quite figure out how that's done in the native `INSERT` implementation.

**Reply by zzzeek** [reply]
running two statements like that would require the use of event handlers like before_execute() or before_cursor_execute().  However, the value of the sequence execution would not be received by the ORM which relies upon https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=cursorresult#sqlalchemy.engine.BaseCursorResult.inserted_primary_key to get back the new value.

what you're asking here is again the same question as before with manipulating the `SQLCompiler` state such that DefaultDialect sees a default value that needs to be "pre-executed".  I can make an example but I cannot guarantee it isn't going to break across major releases.

Basically as long as the compiler is given an insert().from_select(), it will know that "pre executing a sequence" is not appropriate, because SELECT is against any number of rows.    Therefore the correct place for this very specific snowflake translation to occur is when rendering the string form of INSERT, and not by changing the kind of insert() construct that the compiler is receiving.   The snowflake dialect itself should be doing this string translation in its `visit_insert` method.

**Reply by luciengaitskell** [reply]
Got it, thanks again for that explanation! I'll leave that alone for now, but good to know 🙌 .
